File: src/main/java/cat/mvmike/numbertotext/magnitude/Units.java
package cat.mvmike.numbertotext.magnitude;

import static cat.mvmike.numbertotext.language.Constants.EMPTY;
import static cat.mvmike.numbertotext.language.Constants.PLURAL;
import static cat.mvmike.numbertotext.language.Constants.SPACE;
import static cat.mvmike.numbertotext.util.StringUtils.isEmpty;

public final class Units extends BaseMagnitude {

    private final boolean includeZero;

    public Units(int number) {
        super(number);
        this.includeZero = false;
    }

    public Units(int number, boolean includeZero) {
        super(number);
        this.includeZero = includeZero;
    }

    @Override
    public String get() {
        String tens = (new Hundreds(number).isMultipleOf() && !includeZero ? EMPTY : new Tens(number % 100).get());
        String hundreds = new Hundreds((number / 100) % 10).get();

        return isEmpty(hundreds) ? tens :
                isEmpty(tens) ? hundreds : hundreds + SPACE + tens;
    }

    @Override
    public String getCurrency(String currency) {
        return isEmpty(currency) ? EMPTY : SPACE + currency + (number % 10 != 1 ? PLURAL : EMPTY);
    }
}

File: src/main/kotlin/cat/mvmike/numbertotext/magnitude/Units.kt
// Copyright (c) 2016, Miquel Martí <miquelmarti111@gmail.com>
// See LICENSE for licensing information
package cat.mvmike.numbertotext.magnitude

import cat.mvmike.numbertotext.language.Constants.EMPTY
import cat.mvmike.numbertotext.language.Constants.PLURAL
import cat.mvmike.numbertotext.language.Constants.SPACE
import cat.mvmike.numbertotext.language.NumberLiteral.N_1
import cat.mvmike.numbertotext.language.NumberLiteral.N_100

class Units(
    private val number: Int,
    private val includeZero: Boolean = false
) {

    fun get(): String {
        val tens = when {
            number.isMultipleOfHundreds() && !includeZero -> EMPTY
            else -> Tens(number.modOfHundred()).get()
        }
        val hundreds = Hundreds((number / N_100.value).modOfTen()).get()

        return when {
            hundreds.isEmpty() -> tens
            tens.isEmpty() -> hundreds
            else -> hundreds + SPACE + tens
        }
    }

    fun getCurrency(currency: String?) = when {
        currency.isNullOrEmpty() -> EMPTY
        else -> SPACE + currency + addPluralIfNeeded()
    }

    private fun addPluralIfNeeded() = when (N_1.value) {
        number.modOfTen() -> EMPTY
        else -> PLURAL
    }
}